cql3/restrictions/statement_restrictions.cc

```diff
@@ -1057,6 +1061,359 @@ ::shared_ptr<expression> make_conjunction(::shared_ptr<expression> a, ::shared_p
+    if (!rhs) {
+        return false;
+    }
+    const auto value = get_value(lhs, selection, cells, options);
+    if (!value) {
+        return false;
+    }
+    return comparator(lhs)->equal(*value, *rhs);
+}
+
+/// True iff columns' values equal t.
+bool equal(::shared_ptr<term> t, const std::vector<column_value>& columns, const selection& selection,
+           row_data& cells, const query_options& options) {
+    if (columns.size() > 1) {
+        auto multi = dynamic_pointer_cast<multi_item_terminal>(t);
+        if (!multi) {
+            throw std::logic_error("RHS for multi-column is not a tuple");
+        }
+        const auto& rhs = multi->get_elements();
+        if (__builtin_expect(rhs.size() != columns.size(), false)) {
+            throw std::logic_error("LHS and RHS size mismatch");
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            if (!equal(rhs[i], columns[i], selection, cells, options)) {
+                return false;
+            }
+        }
+        return true;
+    } else if (columns.size() == 1) {
+        return equal(to_bytes_opt(t->bind_and_get(options)), columns[0], selection, cells, options);
+    } else {
+        throw std::logic_error("empty tuple on LHS of =");
+    }
+}
+
+/// True iff lhs is limited by rhs in the manner prescribed by op.
+bool limits(const bytes& lhs, const operator_type& op, const bytes& rhs, const abstract_type& type) {
+    if (!op.is_compare()) {
+        throw std::logic_error("limits() called on non-compare op");
+    }
+    const auto cmp = type.as_tri_comparator()(lhs, rhs);
+    if (cmp < 0) {
+        return op == operator_type::LT || op == operator_type::LTE || op == operator_type::NEQ;
+    } else if (cmp > 0) {
+        return op == operator_type::GT || op == operator_type::GTE || op == operator_type::NEQ;
+    } else {
+        return op == operator_type::LTE || op == operator_type::GTE || op == operator_type::EQ;
+    }
+}
+
+/// True iff the value of opr.lhs is limited by opr.rhs in the manner prescribed by opr.op.
+bool limits(const binary_operator& opr, const selection& selection, row_data& cells,
+            const query_options& options) {
+    if (!opr.op.is_slice()) { // For EQ or NEQ, use equals().
+        throw std::logic_error("limits() called on non-slice op");
+    }
+    const auto& columns = std::get<0>(opr.lhs);
+    if (columns.size() > 1) {
+        auto multi = dynamic_pointer_cast<multi_item_terminal>(opr.rhs);
+        if (!multi) {
+            throw std::logic_error("RHS for multi-column is not a tuple");
+        }
+        const auto& rhs = multi->get_elements();
+        if (rhs.size() != columns.size()) {
+            throw std::logic_error("LHS and RHS size mismatch");
+        }
+        for (size_t i = 0; i < rhs.size(); ++i) {
+            const auto cmp = comparator(columns[i])->as_tri_comparator()(
+                    *get_value(columns[i], selection, cells, options),
```

Comment:
Why is it safe to extract the value from get_value()?

Reply:
I guess it's because multi terminal compares are always for the partition key? But it's not clear here.

Reply:
No, tihis is for the clustering key. Either way, clarification is needed.

Reply:
Clarified with a comment.